src/core/public/application/utils/parse_app_url.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { getUrlOrigin } from '@kbn/std';
+import { resolve } from 'url';
+import { IBasePath } from '../../http';
+import { App, ParsedAppUrl } from '../types';
+
+/**
+ * Parse given url and return the associated app id and path if any app matches, or undefined if none do.
+ * Input can either be:
+ *
+ * - a absolute path containing the basePath,
+ *   e.g `/base-path/app/my-app/some-path`
+ *
+ * - an absolute url matching the `origin` of the kibana instance (as seen by the browser),
+ *   e.g `https://kibana:8080/base-path/app/my-app/some-path`
+ *
+ * - a path relative to the provided `currentUrl`.
+ *   e.g with `currentUrl` being `https://kibana:8080/base-path/app/current-app/some-path`
+ *   `../other-app/other-path` will be converted to `/base-path/app/other-app/other-path`
+ */
+export const parseAppUrl = (
+  url: string,
+  basePath: IBasePath,
+  apps: Map<string, App<unknown>>,
+  currentUrl: string = window.location.href
+): ParsedAppUrl | undefined => {
+  const currentOrigin = getUrlOrigin(currentUrl);
+  if (!currentOrigin) {
```

Comment:
The actual diff for the changes on `parseAppUrl` can be seen in this commit: https://github.com/elastic/kibana/pull/78565/commits/e14858fffa76d4077f2e378e965ccad960195887